src/Psysh/Drupal8ForkingLoop.php

```diff
@@ -0,0 +1,167 @@
+      }
+
+      $content = stream_get_contents($down);
+      fclose($down);
+
+      if ($content) {
+        $shell->setScopeVariables(@unserialize($content));
+      }
+
+      return;
+    }
+
+    // This is the child process. It's going to do all the work.
+    if (function_exists('setproctitle')) {
+      setproctitle('psysh (loop)');
+    }
+
+    // We won't be needing this one.
+    fclose($down);
+
+    // Let's do some processing.
+    parent::run($shell);
+
+    // Send the scope variables back up to the main thread
+    fwrite($up, $this->serializeReturn($shell->getScopeVariables()));
+    fclose($up);
+
+    posix_kill(posix_getpid(), SIGKILL);
+  }
+
+  /**
+   * Create a savegame at the start of each loop iteration.
+   */
+  public function beforeLoop() {
+    $this->createSavegame();
+  }
+
+  /**
+   * Clean up old savegames at the end of each loop iteration.
+   */
+  public function afterLoop() {
+    // if there's an old savegame hanging around, let's kill it.
+    if (isset($this->savegame)) {
+      posix_kill($this->savegame, SIGKILL);
+      pcntl_signal_dispatch();
+    }
+  }
+
+  /**
+   * Create a savegame fork.
+   *
+   * The savegame contains the current execution state, and can be resumed in
+   * the event that the worker dies unexpectedly (for example, by encountering
+   * a PHP fatal error).
+   */
+  private function createSavegame() {
+    // the current process will become the savegame
+    $this->savegame = posix_getpid();
+
+    $pid = pcntl_fork();
+    if ($pid < 0) {
+      throw new \RuntimeException('Unable to create savegame fork.');
+    } elseif ($pid > 0) {
+      // we're the savegame now... let's wait and see what happens
+      pcntl_waitpid($pid, $status);
+
+      // worker exited cleanly, let's bail
+      if (!pcntl_wexitstatus($status)) {
+        posix_kill(posix_getpid(), SIGKILL);
+      }
+
+      // worker didn't exit cleanly, we'll need to have another go
+      $this->createSavegame();
+    }
+  }
+
+  /**
+   * Serialize all serializable return values.
+   *
+   * A naïve serialization will run into issues if there is a Closure or
+   * SimpleXMLElement (among other things) in scope when exiting the execution
+   * loop. We'll just ignore these unserializable classes, and serialize what
+   * we can.
+   *
+   * @param array $return
+   *
+   * @return string
+   */
+  private function serializeReturn(array $return) {
+    $serializable = array();
+
+    foreach ($return as $key => $value) {
+      // No need to return magic variables
+      if ($key === '_' || $key === '_e') {
+        continue;
+      }
+
+      // Resources don't error, but they don't serialize well either.
+      if (is_resource($value) || $value instanceof \Closure) {
+        continue;
+      }
+
+      try {
+        @serialize($value);
+        $serializable[$key] = $value;
+      } catch (\Exception $e) {
+        // we'll just ignore this one...
+      }
+      catch (\AssertionError $e) {
```

Comment:
Yeah, I thought about that first, but Psysh supports PHP5.6 and above. So we would need to add some \AssertionError exception like Drupal does. Which I think might be a no go... at least until it was only support PHP7.